Map Staging Area

sjperkins opened the pull request:
I liked the idea of StagingArea but I wanted more functionality out of it. I've implemented an Associative StagingArea backed by C++ `std::map` and `std::unordered_map`.

Functionality:
- Keys are int64, Values are just like StagingArea Tensors or list of Tensors.
- put(key, value), get(key),  pop(key), popitem(), size(), clear() operations
- Can choose ordered `std::map` (Get a basic priority queue with popitem()) or unordered `std::unordered`.
- Can specify a capacity which inserts will wait on, if not, it will be unbounded.

Have tested that this works on CPU + GPU in a custom op library. I haven't tested the compile on my laptop due to #9634, but I've cut and pasted the relevant sections into the tensorflow code base for your perusal.

Let me know if you're keen to merge this and I can clean it up a bit further.

sjperkins committed:
Map Staging Area Operator

tensorflow-jenkins commented:
Can one of the admins verify this patch?

ekelsen commented:
Hi Simon,

Thanks for this nice patch.  We had a couple of thoughts for a re-organization to avoid creating multiple classes with slightly different functionality and naming conventions:

1) Rather than creating new python high level ops, could we slightly modify the existing StagingArea interface by adding an optional key argument to get and put?  If the key is present, then your C++ ops would be used, otherwise the current C++ ops would be used.  The constructor could take an argument that decides on ordering.

2) Your pop method would map to get and if you want to retain the ability to inspect the StagingArea without removing an element, then add a new method 'peek'.

3) You can choose to add fixed capacity to the existing StagingArea or remove it from the new classes (so only unbounded).

Thanks,
Erich

sjperkins commented:
> Rather than creating new python high level ops, could we slightly modify the existing StagingArea interface by adding an optional key argument to get and put? If the key is present, then your C++ ops would be used, otherwise the current C++ ops would be used. The constructor could take an argument that decides on ordering.

I disagree with this approach because two very different data structures are being exposed and I think it would be a mistake to combine their interface. They offer different [affordances](https://en.wikipedia.org/wiki/Affordance) (see Norman's [The Design of Everyday Things](https://en.wikipedia.org/wiki/The_Design_of_Everyday_Things)) to the user. Thus:

1. Mixing the interfaces will be confusing for the end user. StagingArea behaves like a list while MapStagingArea behaves like a dict. 
2. I actually think it will be more difficult to maintain since the combination of interfaces is confusing.
3. Does the data for the list exist in a different container from the dict?
4. What happens if the user puts without a key and then then tries to get with a key?

I think the following would be better:

1. Keep the separate list (StagingArea) vs map (MapStagingArea) python classes
2. Move common functionality into a Base Class (coloc op, dtype, shape, name, return value handling etc.)
3. Rename `get(key)` to `peek(key)` in the dict. I dislike losing the name of the python dict method, but it probably too late to change `get()` to `pop()` in the list.
4. Add `peek(index)`, `size` and `clear` methods to the list (Already got a size op for the list).

> You can choose to add fixed capacity to the existing StagingArea or remove it from the new classes (so only unbounded).

OK I can do this.

ekelsen commented:
That works, thanks again.

Re: 3) Make sure to also change pop in the dict to get.

sjperkins commented:
> Re: 3) Make sure to also change pop in the dict to get.

Just realised it may make more sense `popitem()` changed to `get()`, since `popitem`  currently takes no `key` argument (it returns a random element if unordered or the max element if ordered), while `pop(key)` returns an element based on a key. Do you agree?

I also want to extend the dict methods as follows:

- `peek(key, default=None)`
- `pop(key, default=None)`
- setdefault(key, value, default=None)

where the specified default is returned if no element associated with the key is found.

Also I remember someone (maybe @girving) mentioning somewhere that there's a way of placing an int32 input for a GPU op on the CPU (not through HostMemory). IIRC its not really supported but forcing the int64 key to the CPU is a good case for it. Could someone point me to code that does this?

ekelsen commented:
Maybe if popitem(key=None) == pop() just popitem -> get and "get" rid of pop ?

I don't think is what you want, since it does use HostMemory: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/split_v_op.cc#L383
but I'm not sure what mechanism @girving was thinking of.

Why not do it this way?  You could force the int64 case only to the CPU.

sjperkins committed:
Add map staging operators to the BUILD

sjperkins committed:
Don't be too clever about notifying waiting thread

sjperkins committed:
Add Staging Area Size op

sjperkins committed:
Use predicate form of condition_variable.wait

sjperkins committed:
Add a Staging Area clear op

sjperkins committed:
Add Peek op and capacity to the Staging Area

sjperkins committed:
map_stage_op.cpp -> map_stage_op.cc

sjperkins committed:
size should be int32

sjperkins committed:
Small improvements to staging test case

sjperkins committed:
Create BaseStagingArea

Containing common functionality to StagingArea and BaseStagingArea

sjperkins committed:
Add MapStagingArea test cases

yaroslavvb commented:
Hm, are integer types even supported on GPU? There's [this line](https://github.com/tensorflow/tensorflow/blob/a304537954a865752ad1b18461e6bd67b36082db/tensorflow/core/framework/register_types.h#L171) in register_types

```
// Call "m" on all types supported on GPU.
#define TF_CALL_GPU_NUMBER_TYPES(m) \
  TF_CALL_half(m) TF_CALL_float(m) TF_CALL_double(m)

```

sjperkins committed:
Add an ordering test

sjperkins committed:
Rename get() to peek()

sjperkins committed:
Combine pop(key) and popitem() into get(key=None)

sjperkins committed:
Block on peek methods

Wait for the requested element to be inserted into the staging area.

sjperkins commented:
OK I think this is ready for review.

Please take note of https://github.com/tensorflow/tensorflow/pull/9686/commits/92d4d732cab8c6b8db83a38166298ed9a4e8961c. I've made StagingArea.peek(index), MapStagingArea.peek(key) and MapStagingArea.get(key) block until the requested index/key is present in the staging area. This means that bounds checks/key lookups may frequently occur in waiting getters when inserters insert their data and notify the getters.

ekelsen commented on tensorflow/core/kernels/stage_op.cc:
```diff
@@ -30,26 +30,82 @@ namespace {
+    mutex_lock l(mu_);
+
+    // Wait if the requested index is not available
+    non_empty_cond_var_.wait(l, [&, this]() {
```
[&, this] might redundant

"[this] captures the current object (*this) by reference
[&] captures all automatic variables used in the body of the lambda by reference and current object by reference if exists"

sjperkins replied:
It needs to capture both `index ` and `this`. I've changed the capture of `index` to capture by value as its just an int.

https://github.com/tensorflow/tensorflow/pull/9686/commits/a90c2e59a37acd2dd54d838261831aab34de3474

ekelsen commented on tensorflow/core/kernels/map_stage_op.cc:
```diff
@@ -0,0 +1,497 @@
+public:
+  // public methods
+  explicit StagingMap(int capacity)
+    { capacity_ = capacity; }
```
initialization list

sjperkins replied:
https://github.com/tensorflow/tensorflow/pull/9686/commits/4f8b0a6a6c27c0005c3b34760606cdb2fa81cdd3 and https://github.com/tensorflow/tensorflow/pull/9686/commits/d97bce1c3d100f4f6070890202f2a427cb04616e

ekelsen commented on tensorflow/core/kernels/stage_op.cc:
```diff
@@ -30,26 +30,82 @@ namespace {
+  }
+
+  bool IsFull() {
+    return buf_.size() >= capacity_;
```
If Size grabs the mutex, should IsFull?

sjperkins replied:
Original intention was a private helper method. Made it private.

https://github.com/tensorflow/tensorflow/pull/9686/commits/5ad80b2d5013cb521dd5fc176852748d8c6f8e89

ekelsen commented on tensorflow/core/kernels/map_stage_op.cc:
```diff
@@ -0,0 +1,497 @@
+#ifdef TENSORFLOW_USE_SYCL
+REGISTER_KERNEL_BUILDER(Name("MapStage").HostMemory("key")
+                      .Device(DEVICE_SYCL), MapStageOp<false>);
+REGISTER_KERNEL_BUILDER(Name("MapStage").HostMemory("key")
```
OrderedMapStage

sjperkins replied:
https://github.com/tensorflow/tensorflow/pull/9686/commits/895be638df52a50de74e6f99dbabb5c866df5b95

ekelsen reviewed (commented):
In general, looks very nice, thank you.

sjperkins committed:
Use constructor initializer lists

sjperkins committed:
Arrange class typedefs, members and methods

Also make previously public helper methods like HasBoundedCapacity() and
IsFull() private.

sjperkins committed:
Better docs for StagingArea.clear()

sjperkins committed:
SYCL_DEVICE MapStage => OrderedMapStage

sjperkins committed:
Initializer list fix

sjperkins committed:
device('/gpu:0') => test.gpu_device_name()

sjperkins committed:
Copyright year to 2017

sjperkins committed:
Reverse keys before random shuffle

sjperkins committed:
Python docstring updates

sjperkins committed:
Fix stage op docs

sjperkins committed:
Order RecordInput correctly

sjperkins committed:
Updated MapStage op docs

sjperkins commented on tensorflow/core/ops/data_flow_ops.cc:
```diff
@@ -1963,6 +1965,39 @@ dequeue with many fewer capabilities and options.  This Op is optimized for
+    .SetShapeFn(shape_inference::UnknownShape)
+    .SetIsStateful()
+    .Doc(R"doc(
+Op is similar to a lightweight Dequeue.  The basic funtionality is similar to
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/a9a6ccb43f069cda06f43976be55d35ab95a6d09

sjperkins commented on tensorflow/core/ops/data_flow_ops.cc:
```diff
@@ -1963,6 +1965,39 @@ dequeue with many fewer capabilities and options.  This Op is optimized for
+    .Attr("shared_name: string = ''")
+    .SetShapeFn(shape_inference::ScalarShape)
+    .SetIsStateful()
+    .Doc(R"doc(Staging Area Size)doc");
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/a9a6ccb43f069cda06f43976be55d35ab95a6d09

sjperkins commented on tensorflow/core/ops/data_flow_ops.cc:
```diff
@@ -1963,6 +1965,39 @@ dequeue with many fewer capabilities and options.  This Op is optimized for
+    .Attr("shared_name: string = ''")
+    .SetShapeFn(shape_inference::UnknownShape)
+    .SetIsStateful()
+    .Doc(R"doc(Staging Area Clear)doc");
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/a9a6ccb43f069cda06f43976be55d35ab95a6d09

sjperkins commented on tensorflow/core/ops/data_flow_ops.cc:
```diff
@@ -1986,4 +2021,129 @@ file_parallelism: How many sstables are opened and concurrently iterated over.
+    .Attr("shared_name: string = ''")
+    .SetShapeFn(tensorflow::shape_inference::NoOutputs)
+    .SetIsStateful()
+    .Doc(R"doc(MapStage)doc");
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/a9a6ccb43f069cda06f43976be55d35ab95a6d09

sjperkins commented on tensorflow/python/kernel_tests/map_stage_op_test.py:
```diff
@@ -0,0 +1,244 @@
+# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/f3e0ed97f79fbcb8ac1123761d75295230ae177e

sjperkins commented on tensorflow/python/kernel_tests/map_stage_op_test.py:
```diff
@@ -0,0 +1,244 @@
+      pi = array_ops.placeholder(dtypes.int64)
+      gi = array_ops.placeholder(dtypes.int64)
+      v = 2. * (array_ops.zeros([128, 128]) + x)
+    with ops.device(test.gpu_device_name()):
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/0693e69213e560be8b790ab29dd201769adc7b09

sjperkins commented on tensorflow/python/kernel_tests/map_stage_op_test.py:
```diff
@@ -0,0 +1,244 @@
+
+      # Shuffle them for random insert
+      shuffle_keys = [i for i in keys]
+      random.shuffle(shuffle_keys)
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/ff0c599ed31977df722808d36531548944e35fb6

ekelsen replied:
I think that this still has a small but finite probability of ending up with an unsorted list.  The shuffle could undo the reverse.  Just reverse.  Then you guarantee that the list is not sorted.

sjperkins commented on tensorflow/python/ops/data_flow_ops.py:
```diff
@@ -1584,19 +1642,270 @@ def get(self, name=None):
+                        capacity=self._capacity)
+
+  def clear(self, name=None):
+    """Empties the staging area."""
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/1cd58f2c241676ce933a15beb680cfc0f53a795c

sjperkins commented on tensorflow/python/ops/data_flow_ops.py:
```diff
@@ -1584,19 +1642,270 @@ def get(self, name=None):
+class MapStagingArea(BaseStagingArea):
+    """
+    Class for staging inputs. Similar to `StagingArea` but behaves
+    like a hashtable with support for puts, gets, pops, popitems,
```
https://github.com/tensorflow/tensorflow/pull/9686/commits/ce326f6aa03e4375331a8ddbcec25e0dc97c681b

sjperkins committed:
Capture index by value instead of reference

Its just an int.

sjperkins commented:
Just had a thought that it might be good to move the C++ StagingArea and MapStagingArea classes in separate header and source files so that they can be reused as primitives in other ops. I'll try do this soon.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

sjperkins committed:
Add support for Staging Area memory limit

Mostly useful for constraining the amount of memory used for Staging
Areas placed on the GPU.

sjperkins committed:
Use key Tensor as key rather than int64

sjperkins committed:
Pass dtypes to StagingMap

sjperkins committed:
CamelCase types

sjperkins committed:
Support barrier operation on the MapStagingArea

- It is now possible to insert incomplete elements.
- Key std::map and std::unordered_map on scalar Tensor int64
  (avoids unnecessary allocations)

sjperkins commented:
One question. I see that structures like FIFOQueue use PersistentTensors, presumably for reasons suggested [here](https://github.com/tensorflow/tensorflow/blob/554b57f74a99ea18baf616ae5e3cff8b137430e6/tensorflow/core/framework/op_kernel.h#L193-L198):

```c++
// Wraps a tensor that is held by an Op across calls to Compute(). For
// memory safety when using asynchronous devices like GPUs, the system
// must be notified when a Tensor is used inside an Op execution. The
// wrapper ensures that all uses of the Tensor are tracked, because in
// order to retrieve the Tensor the caller must use AccessTensor which
// notifies the context.
class PersistentTensor { ... };
```

Should the tensors in the Staging Area be wrapped in PersistentTensors?

rmlarsen commented:
@ekelsen could you please take another look and  address @sjperkins's questions?

ekelsen commented on tensorflow/core/kernels/map_stage_op.cc:
```diff
@@ -65,6 +65,8 @@ class StagingMap : public ResourceBase
 private:
   // Private variables
   int capacity_;
+  int memory_limit_;
```
When dealing with memory sizes, it's probably best to use size_t.  A signed 32-bit integer could easily be exceed.

sjperkins committed:
Remove shuffling to make test deterministic

sjperkins committed:
Use size_t for indices and positive quantities

ekelsen commented:
Regarding using PersistentTensor, I think PersistentTensor is poorly named and that that comment is a bit out-of-date.  Just using Tensors should be fine.  The PersistentTensor existed to deal with synchronization between multiple GPU streams.

ekelsen reviewed (approved):
LGTM

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@sjperkins Can you address the lint and BUILD file formatting errors here:

https://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/4336/consoleFull

sjperkins committed:
Fix lint errors

- autopep8 data_flow_ops.py indentation to 2 spaces.
- Alphabetically order data_flow lib dependencies in BUILD

sjperkins commented:
Fixed

rmlarsen commented:
@sjperkins Thanks!
@tensorflow-jenkins test this please

rmlarsen commented:
@sjperkins it looks like you missed a few.

sjperkins committed:
extra capacity() should be memory_limit()

sjperkins committed:
Run buildifier

sjperkins commented:
@rmlarsen I ran buildifier on the source tree, but it different extra line (2306) than that mentioned in the log (2355 below). 

```bash
=== Sanity check step 3 of 8: do_buildifier (buildifier check) ===

Running do_buildifier on 287 files

tensorflow/python/kernel_tests/BUILD # reformat 

buildifier took 0 s

FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
2355d2354
< 
Please fix manually or run buildifier <file> to auto-fix.
```

Would you test again please?

ekelsen commented:
BTW, can you expand on your use case for the map version of the staging area?  Just out of curiosity...

rmlarsen commented:
@tensorflow-jenkins test this please

sjperkins commented:
Taking another look

sjperkins committed:
Fix OrderedMapIncompleteSize kernel name

sjperkins committed:
Fix device name in colocation test

Device name is "" if no GPU is present.

sjperkins committed:
Fix subtle size_t to int conversion bug

sjperkins commented:
> BTW, can you expand on your use case for the map version of the staging area? Just out of curiosity...

1. Partial inserts. Feeding one Staging Area (key, tuple) from multiple (possibly, distributed) locations. i.e. scatter portions of the problem from master/chief and feed other portions of the problem on a local worker. I've a [data source](https://github.com/ska-sa/montblanc/blob/75e47e5a4ee8e7f599b6e5805333a39754beadd8/montblanc/impl/rime/tensorflow/sources/ms_source_provider.py#L84-L213) interface to my computation and I want to allow my users to feed data in multiple places.
2. Capacity and Memory Budget. Feeding a potentially very large set of Tensors onto the GPU asynchronously and making sure that only a restricted portion of the set is on the Staging Area at one time.
3. Peek functionality. Caching Tensors on GPU for the duration of other operations (i.e. point 2).
4. Ordering. I'm grouping data (and computation) by a integer key defined by data tile extents. There's a partial ordering here thats useful to me. Ultimately I want to use this key for [DHT](https://en.wikipedia.org/wiki/Distributed_hash_table) functionality, but a priority queue is always useful.
5. Holding ref-counted Tensors, rather than memcpying tensors slices around.

In general, having a stateful hash table in the Graph, logically grouping ref-counted Tensor objects together on either CPUs or GPUs is pretty useful. The Queues always seemed heavyweight (https://github.com/tensorflow/tensorflow/issues/3009, https://github.com/tensorflow/tensorflow/issues/5907, https://github.com/tensorflow/tensorflow/issues/6845) so I wanted to avoid memcpys. I saw some tensor slicing memcpys in Barrier, and it's not always clear what's happening lower down in the PriorityQueue/FIFOQueue level.

rmlarsen commented:
@tensorflow-jenkins test this please

sjperkins committed:
Add #include <numeric> for windows builds

rmlarsen commented:
@tensorflow-jenkins test this please

sjperkins committed:
Add #include <numeric> in stage_op.cc too

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
Thanks for the contribution. This is pretty cool stuff!

sjperkins commented:
@rmlarsen @ekelsen Thanks for the reviews and accepting the PR.

ekelsen commented:
Ditto on the thanks - this is a great contribution.

sjperkins commented:
Will this be included in the 1.2.0 release?

ekelsen commented:
1.2 seems likely, but I can't say for sure.

But one minor request - the staging_op_test now takes a bit too long for a "small" test.  Can you submit a minor PR that bumps the test from small -> medium?  I'll review and approve quickly.  Thanks!

ekelsen commented:
Could also consider making the test have a shard count of 2. "shard_count = 2,"

sjperkins commented:
> 1.2 seems likely, but I can't say for sure.

It would be helpful if it made it in, as I'm developing with the Map Staging Area as a dependency.

ekelsen commented:
It seems that this PR has caused the staging_op_test to become flaky.  Sometimes it passes in 30s, sometimes it fails and times out in 300s.  My original intuition to increase the test time was incorrect.  Can you look into a possible race condition?

ekelsen commented:
(This is blocking merging now, so it is rather important to fix the root cause of the hanging test.)

jhseu commented:
We'll likely revert this change for now to unblock merging. Feel free to resend the pull request with the fix.

sjperkins commented:
@jhseu I suspect the fix is in #10276 as I've done 5000 runs on `stage_op_test` and have seen no errors. `map_stage_op_test` is on 572/5000 runs on  now.
